Enable work variant firewall with iptables-compat

Izorkin opened the pull request:
###### Motivation for this change
Enable work variant firewall and fail2ban with iptables-compat - nftables compatibility.

Example configuration:
```
  services.fail2ban = {
    enable = true;
    packageFirewall = pkgs.iptables-compat;
    jails = {
      DEFAULT = ''
        enabled  = true
      '';
      ssh-iptables = ''
        enabled  = true
      '';
    };
  };

  networking.firewall.enable = true;
  networking.firewall.package = pkgs.iptables-compat;
```
Work tested in a virtual machine.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

Izorkin committed:
iptables-compat: init iptables with nftables compatibility

Izorkin committed:
nixos/firewall: add package option

Izorkin commented:
@GrahamcOfBorg build iptables iptables-compat

aanderse commented on nixos/modules/services/security/fail2ban.nix:
```diff
@@ -31,21 +31,35 @@ in
+        description = "The fail2ban package to use for running the fail2ban service.";
+      };
+
+      packageFirewall = mkOption {
```
Do we actually need a `package` option for the firewall here given we already have `networking.firewall.package` option? I would have thought that using `config.networking.firewall.package` in this module would be enough.

Izorkin replied:
Without this parameter fail2ban will access the package pkgs.iptables directly. And we need to use pkgs.iptables-compat. Also then use pkgs.nftables (need change actions to nftables-multiport)

aanderse replied:
If the `firewall` module uses `config.networking.firewall.package` then you could make it use `pkgs.iptables-compat`. I'm not sure I follow.

Izorkin replied:
To change the fail2ban service modification later, I will need the packageFirewall option.
Example my variant fail2ban (corrections required) - https://github.com/Izorkin/nur-packages/blob/master/modules/services/fail2ban.nix

aanderse reviewed (changes requested):
@Izorkin overall looks like a good change :+1: I don't know enough about the networking stack to merge this, but we can ping a few people and try to get a good thorough review+merge.

FRidh commented on pkgs/os-specific/linux/iptables/default.nix:
```diff
@@ -1,32 +1,49 @@
-  pname = "iptables";
   version = "1.8.3";
+  pname = "iptables";
+  name = pname + "-" + (if modeCompat then "compat" else "legacy") + "-" + version;
```
I don't see the value of changing the name.

Izorkin replied:
Fixed.

FRidh commented on pkgs/top-level/all-packages.nix:
```diff
@@ -15485,7 +15485,9 @@ in
-  iptables = callPackage ../os-specific/linux/iptables { };
+  iptables = iptables-legacy;
+  iptables-legacy = callPackage ../os-specific/linux/iptables { };
+  iptables-compat = callPackage ../os-specific/linux/iptables { modeCompat = true; };
```
Is it worth building this variant?

Izorkin replied:
@FRidh i set the legacy default mode of operation. Then can switch to a new compat mode.
Works normal on my home server.
Or leave only compat mode?

Izorkin replied:
On OS archlinux, debian 10 already is used iptables-compat

Mic92 replied:
I would just call this variant `iptables-nft` or `iptables-nftables` for clarity.

Izorkin replied:
In the OS Debian distribution, the package is also named iptables-compat
https://manpages.debian.org/testing/iptables-nftables-compat/iptables-compat.8.en.html

Izorkin replied:
Need to enable use iptables-compat to default?

FRidh reviewed (changes requested):
The package options are good I think. Not sure about always building the variant nor renaming the package. Can't judge the rest of the PR.

Izorkin commented:
Updated default configuration in the serice fail2ban. Added custom options: banaction, bantime-increment, bantime-increment-config.
The bantime-increment option is for version fail2ban 0.11
@aanderse please review update.

aanderse commented:
@Izorkin you should ask @FRidh and those more familiar with this module approve this PR. Unfortunately I don't know enough to say yes.

Izorkin commented:
Moved fail2ban changes to PR #67931

Mic92 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -15629,7 +15629,9 @@ in
 
-  iptables = callPackage ../os-specific/linux/iptables { };
+  iptables = iptables-legacy;
+  iptables-legacy = callPackage ../os-specific/linux/iptables { };
```
I would not call it `legacy` because the current iptables seems no where near deprecation at the moment.

Izorkin replied:
The Netfilter project, has agreed to consider iptables as a legacy tool.
https://ral-arturo.org/2018/06/16/nfws2018.html

Mic92 replied:
Fair enough.

Mic92 commented on nixos/modules/services/networking/firewall.nix:
```diff
@@ -331,6 +331,16 @@ in
+      package = mkOption {
+        type = types.package;
+        default = pkgs.iptables;
+        example = pkgs.iptables-compat;
```
I think we can reconsider using `iptables-nft` by default if `nftables` is enabled.

Mic92 committed:
iptables-compat: rename to iptables-nftables-compat

Same as debian, more clarity what the package is about.

Mic92 committed:
nixos/firewall: rename iptables-compat to iptables-nftables-compat

Izorkin commented:
@Mic92 thanks!
